### PR TITLE
Few changes done in the images preprocessing part

### PR DESCRIPTION
**Pull request by ppujol76**
I changed the normalization transform of the images to match the Imagenet means and variances. Also changed the IMAGE_SIZE to 224 in the hparam in json file in order to match with the requirements for the pretrained nets.

### CODE REVIEW

**Review by lucaswerner90 — COMMENTED** [review]
looks ok to me, have you tried running the train.py?

**Comment by ppujol76** [comment]
Yes, and it works fine

**Review by lucaswerner90 — APPROVED** [review]

File: dataset/main.py [hunk 1]
```diff
@@ -56,6 +85,7 @@ def __getitem__(self,idx):
             string:			Caption of the image
         """
         filename, caption = self.dataframe.iloc[idx,:]
+        print(filename, caption)
```
Comment (lucaswerner90) [inline dataset/main.py:88]: remove this

File: dataset/main.py [hunk 2]
```diff
@@ -15,12 +15,41 @@ class Flickr8kDataset(Dataset):
...
         self.images_folder = os.path.join(dataset_folder,'Images')
+        
         self.dataframe = pd.read_csv(open(os.path.join(dataset_folder,'captions.txt'),'r'))
 
+        # Open file with the images for train, test or eval
+        if split=='train':
+            imgs_file = open(os.path.join('data','Flickr_8k.trainImages.txt'), "r")
+        elif split=='test':
+            imgs_file = open(os.path.join('data','Flickr_8k.testImages.txt'), "r")
+        if split=='eval':
```
Comment (lucaswerner90) [inline dataset/main.py:30]: `if split == 'eval'  ` change it for `else:`

File: dataset/main.py [hunk 3]
```diff
@@ -15,12 +15,41 @@ class Flickr8kDataset(Dataset):
...
+        # Convert the dataframe to list
+        llista = list(zip(self.dataframe['image'],self.dataframe['caption']))
+        #  print(len(llista))
+
+        # Create a list with the image and captions of the images requested (train, test or eval)
+        c=[]
+        for a,b in llista:
+            if a in imgs:
+                c.append([a,b])
+        #  print(len(c))
```
Comment (lucaswerner90) [inline dataset/main.py:48]: remove this comment

File: dataset/main.py [hunk 4]
```diff
@@ -15,12 +15,41 @@ class Flickr8kDataset(Dataset):
...
+
+        # Load images file names to a list
+        imgs = []
+        for linea in imgs_file:
+            imgs.append(linea.rstrip('\n'))
+        imgs_file.close()
+            
+        # Convert the dataframe to list
+        llista = list(zip(self.dataframe['image'],self.dataframe['caption']))
+        #  print(len(llista))
```
Comment (lucaswerner90) [inline dataset/main.py:41]: remove this comment

File: main.py [hunk 5]
```diff
@@ -31,8 +31,7 @@ def main():
 		# we should use the ImageNet values to normalize the dataset.
 		# Otherwise we could just normalize the values between -1 and 1 using the 
 		# standard mean and standard deviation
-		# transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225]),
-		transforms.Normalize(mean=[0.5, 0.5, 0.5], std=[0.5, 0.5, 0.5]),
+		transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225]),
```
Comment (lucaswerner90) [inline main.py:34]: could you put these values into the `config.json`?

File: model/encoder.py [hunk 6]
```diff
@@ -1,9 +1,11 @@
 from torch.functional import Tensor
 import torchvision.models as models
 import torch.nn as nn
-class Encoder(nn.Module):
+#import matplotlib.pyplot as plt
```
Comment (lucaswerner90) [inline model/encoder.py:4]: remove this commented import

File: model/encoder.py [hunk 7]
```diff
@@ -1,9 +1,11 @@
...
+
+class Encoder_VGG16(nn.Module):
     def __init__(self):
-        super(Encoder, self).__init__()
+        super(Encoder_VGG16, self).__init__()
         pretrained_model = models.vgg16(pretrained=True)
         self.conv_base = pretrained_model.features
         # Freeze All layers as they will be used for inference
```
Comment (lucaswerner90) [inline model/encoder.py:11]: this comment won't probably belong here

File: model/encoder.py [hunk 8]
```diff
@@ -12,12 +14,64 @@ def __init__(self):
...
+        return features
+
+class Encoder_ResNet50(nn.Module):
+    def __init__(self):
+        super(Encoder_ResNet50, self).__init__()
+        pretrained_model = models.resnet50(pretrained=True)
+
+        modules = list(pretrained_model.children())[:-2]
+        self.conv_base = nn.Sequential(*modules)
+#        print(pretrained_model)
```
Comment (lucaswerner90) [inline model/encoder.py:33]: remove comments

File: model/encoder.py [hunk 9]
```diff
@@ -12,12 +14,64 @@ def __init__(self):
...
+        features = self.flat(features)
+        # For an image size of (224x224) --> features dims  (batch_size, feat_maps=2048, 7x7=49)
+        return features
+
+class Encoder_DenseNet(nn.Module):
+    def __init__(self):
+        super(Encoder_DenseNet, self).__init__()
+        pretrained_model = models.densenet161(pretrained=True)
+        self.conv_base = pretrained_model.features
+#        print(pretrained_model)
```
Comment (lucaswerner90) [inline model/encoder.py:57]: remove comments

File: model/encoder.py [hunk 10]
```diff
@@ -12,12 +14,64 @@ def __init__(self):
...
+        for param in self.conv_base.parameters():
+            param.requires_grad = False
+
+        # Flaten layer that flatten the dimensions 2 and 3 (H and W of the feature maps respectively)
+        self.flat = nn.Flatten(2,3)
         self.relu = nn.ReLU()
 
     def forward(self, x):
-        # (batch_size, feat_maps=512, H=7 , W=7) or (batch_size, 512, 14 , 14)
+        # For an image size of (224x224) --> x dims  (batch_size, 3, H=224 , W=224)
```
Comment (lucaswerner90) [inline model/encoder.py:69]: thanks for this! it's really useful!

**Review by lucaswerner90 — CHANGES_REQUESTED** [review]
It looks good except for my minor comments. Thanks for this Pere!

File: dataset/main.py (hunk 2 above)
Comment (amolero6) [reply dataset/main.py:30]: Do we really need 3 splits? I think we should be ok with 2

File: main.py (hunk 5 above)
Comment (ppujol76) [reply main.py:34]: Sure!

File: model/encoder.py (hunk 9 above)
Comment (ppujol76) [reply model/encoder.py:57]: All comments removed
